### DISCUSSION TITLE
Handle retransmissions to avoid startup errors and/or delays in the library?

### DISCUSSION THREAD

**Discussion by vtrebicky** [Makuna/DFMiniMp3#110] [Q&A]
Hey! Thanks for the great library!

Sometimes I observe `0x40` when calling `getVolume()` right after startup. I assume that's because the module is still reading the SD card. I'm thinking if there's a good way for the library to handle this without calling `delay`.

I've seen that the library puts delays in a couple of places, e.g., in [`setVolume()`](https://github.com/Makuna/DFMiniMp3/blob/13baf94972cbb143d7e996ca3f8f060f949b8f44/src/DFMiniMp3.h#L310). I presume this is to avoid the `0x40` error if `getVolume()` got called immediately after `setVolume()`. I'm thinking the controller should be able to parse the `0x40` response and retransmit if requested.

The ["spec doc"](https://picaxe.com/docs/spe033.pdf) indeed mentions "request retransmission" in the context of the `0x40` response code. In section 3.3.4 it even specifies 3 possible arguments. Unfortunately, I only ever witnessed undocumented arguments in the `getVolume()` case. AFAICT, the original "DFPlayer Mini" uses **4**. I also tested a clone (it says "DFPlayer Mini HW-247A" on it) which uses **3** for the argument.

Would it make sense to try to remove delays from the library and implement retransmissions? I guess it could be a nightmare to rely on undocumented error arguments?

**Comment by Makuna** [comment]
The "original" documents stated the delays and didn't mention anything about retransmission; but that was a long time ago and documents were thin.  I will spend a little time looking at the newest documents and see what is the modern way.

**Reply by vtrebicky** [reply]
Cool, thanks! I think these are the 3 key parts in the doc I know about:

![image](https://user-images.githubusercontent.com/68817231/226274810-80d3a855-95d4-4848-b6fb-b6b9a615c86f.png)
![image](https://user-images.githubusercontent.com/68817231/226274895-ccc4409e-706f-42bc-b6fd-3b9505685fb2.png)
![image](https://user-images.githubusercontent.com/68817231/226274957-d02f5223-436f-47ee-9066-86113609d6f8.png)

Apart from the strange language (to say the least), I'm wondering if there's anything to make out of this, though. There seems to be a variety of hardware, each behaving slightly differently. For example the "original" module uses the `0x3f` code. It sends it both right after successful initialization and as a response to an MCU query. That was surely meant as an elegant way to handle the start delay problem. Alas, at least some of the clones don't do that (like the one I have). And I can't see the library detecting which hardware it's talking to. The version number is moot, it's always set to `0xff`...

**Comment by Makuna** [comment]
One way for the sketch writer to avoid the issue you hit at startup is to wait until the device `isOnline()` before setting the volume.  The examples should be improved to do this; but I think that even that isn't consistent as sometimes it returns only card inserted or card removed and not the (0x3f) Source Online.  (looking at the code, it does set online for all three of these!).

I just checked my old spec and found this
```
. MCU can not send commands to control the module until the initialization of the module is done and a data is
returned, otherwise the commands sent by MCU will be ignored and also this will effect initializing of the module.
```

It is one of the things that happens with the Chinese chips, they revise often, document poorly, and really don't think about backward compatibility.  I think the originals were rushed out with lots of errors but generally functional and they have been improving them over revisions.  I have two modules that are identical (chip numbers/circuit boards) that I bought about 6 months apart and they act differently.  I don't know a way at run time way of requesting versions.  I have never seen anything documented for it.

**Comment by Makuna** [comment]
I have an initial set of changes that use ACK and ERROR messages to retransmit requests until it gets a response.  

As I was worried about, during the time after reset (as the device initializes) it my send notifications (cards online) but it will not respond to requests.  It will neither send ACK nor will it respond with an error.  They just go no where and the library has to rely on the UART timeout as it waits for some sort of "expected" response (valid response, ack, or error).

Still reviewing leaving the ack/error/retry in.

**Comment by Makuna** [comment]
https://github.com/Makuna/DFMiniMp3/pull/112

**Comment by Makuna** [comment]
While investigating this new work, I came across some other inconsistencies I need to get some clarity on with the MH2024K-16SS variant.  Does anyone have one to try something?  Basically, turn on the debug, run the "PlayMp3" example, capture the debugoutput and get that to me.  Specially looking for when a track finishing playing notification details.

**Comment by Makuna** [comment]
https://github.com/Makuna/DFMiniMp3/pull/114
